validation: Remove REJECT code from CValidationState

jnewbery opened the pull request:
We no longer send BIP 61 REJECT messages, so there's no need to set
a REJECT code in the CValidationState object.

Note that there is a minor bug fix in p2p behaviour here. Because the
call to `MaybePunishNode()` in `PeerLogicValidation::BlockChecked()` only
previously happened if the REJECT code was > 0 and < `REJECT_INTERNAL`,
then there are cases were `MaybePunishNode()` can get called where it
wasn't previously:

- when `AcceptBlockHeader()` fails with `CACHED_INVALID`.
- when `AcceptBlockHeader()` fails with `BLOCK_MISSING_PREV`.

Note that `BlockChecked()` cannot fail with an 'internal' reject code. The
only internal reject code was `REJECT_HIGHFEE`, which was only set in
ATMP.

This reverts a minor bug introduced in 5d08c9c579ba8cc7b684105c6a08263992b08d52.

jnewbery committed:
[validation] Fix peer punishment for bad blocks

Because the call to MaybePunishNode() in
PeerLogicValidation::BlockChecked() only previously happened if the
REJECT code was > 0 and < REJECT_INTERNAL, then there are cases were
MaybePunishNode() can get called where it wasn't previously:

- when AcceptBlockHeader() fails with CACHED_INVALID.
- when AcceptBlockHeader() fails with BLOCK_MISSING_PREV.

Note that BlockChecked() cannot fail with an 'internal' reject code. The
only internal reject code was REJECT_HIGHFEE, which was only set in
ATMP.

This change restores the behaviour pre-commit
5d08c9c579ba8cc7b684105c6a08263992b08d52 which did punish nodes that
sent us CACHED_INVALID and BLOCK_MISSING_PREV blocks.

jnewbery committed:
[validation] Remove REJECT code from CValidationState

We no longer send BIP 61 REJECT messages, so there's no need to set
a REJECT code in the CValidationState object.

jnewbery commented:
Builds on #15437. Please review that PR first.

laanwj commented on doc/bips.md:
```diff
@@ -15,7 +15,6 @@ BIPs that are implemented by Bitcoin Core (up-to-date up to **v0.18.0**):
 * [`BIP 35`](https://github.com/bitcoin/bips/blob/master/bip-0035.mediawiki): The 'mempool' protocol message (and the protocol version bump to 60002) has been implemented since **v0.7.0** ([PR #1641](https://github.com/bitcoin/bitcoin/pull/1641)). As of **v0.13.0**, this is only available for `NODE_BLOOM` (BIP 111) peers.
 * [`BIP 37`](https://github.com/bitcoin/bips/blob/master/bip-0037.mediawiki): The bloom filtering for transaction relaying, partial Merkle trees for blocks, and the protocol version bump to 70001 (enabling low-bandwidth SPV clients) has been implemented since **v0.8.0** ([PR #1795](https://github.com/bitcoin/bitcoin/pull/1795)). Disabled by default since **v0.19.0**, can be enabled by the `-peerb ...
 * [`BIP 42`](https://github.com/bitcoin/bips/blob/master/bip-0042.mediawiki): The bug that would have caused the subsidy schedule to resume after block 13440000 was fixed in **v0.9.2** ([PR #3842](https://github.com/bitcoin/bitcoin/pull/3842)).
-* [`BIP 61`](https://github.com/bitcoin/bips/blob/master/bip-0061.mediawiki): The 'reject' protocol message (and the protocol version bump to 70002) was added in **v0.9.0** ([PR #3185](https://github.com/bitcoin/bitcoin/pull/3185)). Starting **v0.17.0**, whether to send reject messages can be configured with the `-enablebip61` option, and support is deprecated as of **v0.18.0**.
```
IMO it'd be more clear, for historical context, to add to this line that support was removed, than to remove this line

jnewbery replied:
This comment is for the base PR (https://github.com/bitcoin/bitcoin/pull/15437). I've copied it there: https://github.com/bitcoin/bitcoin/pull/15437#discussion_r330051814

ryanofsky commented on src/rpc/rawtransaction.cpp:
```diff
@@ -906,7 +906,7 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
     if (!test_accept_res) {
         if (state.IsInvalid()) {
-            result_0.pushKV("reject-reason", strprintf("%i: %s", state.GetRejectCode(), state.GetRejectReason()));
+            result_0.pushKV("reject-reason", strprintf("%s", state.GetRejectReason()));
```
In commit "[validation] Remove REJECT code from CValidationState" (6f8876e155024181a455fc457ae94a38d0da3626)

I don't think it's necessary, but you could potentially mention this change in release notes, in case someone is relying on the RPC string having a certain format.

ryanofsky reviewed (approved):
Code review ACK 6f8876e155024181a455fc457ae94a38d0da3626 (last commit only)

jnewbery committed:
[docs] Add release notes for removal of REJECT reasons

jnewbery commented:
Thanks for the review @ryanofsky . I've updated the release notes as requested.

laanwj commented:
Nice cleanup.

jnewbery commented:
Rebased

ryanofsky reviewed (approved):
Code review ACK 982ff1a424efefcd7403ad1fb2b1206fd5297035. No changes since last review other than rebase and adding release notes

ryanofsky commented:
>  there are cases were `MaybePunishNode()` can get called where it
> wasn't previously

I didn't see this in my previous reviews, and I'm not really sure what the implications are. It might help for the comment to say what the original reason for not calling MaybePunish was, and why it's no longer applicable or important now

MarcoFalke commented:
Was about to ask the same question, thanks @ryanofsky

jnewbery commented:
It looks like the bug was introduced here: https://github.com/bitcoin/bitcoin/pull/10135/files. That PR fixed a bug where we'd send out REJECT messages with a reject code of 0 (not valid in BIP 61), but it also inadvertently stopped us from punishing nodes that sent a block that failed here https://github.com/jnewbery/bitcoin/blob/5d08c9c579ba8cc7b684105c6a08263992b08d52/src/validation.cpp#L3077 and here https://github.com/jnewbery/bitcoin/blob/5d08c9c579ba8cc7b684105c6a08263992b08d52/src/validation.cpp#L3088 where the reject code wasn't set in the `state.Invalid()` call.

This PR restores that old behaviour of maybe punishing peers that send us blocks that fail `CACHED_INVALID` or `BLOCK_MISSING_PREV`. There's already code to handle those failure reasons in `MaybePunishNode()`.

I've updated the commit log with more detail, and added a comment in the net_processing `BlockChecked()` callback.

ryanofsky commented:
Code review ACK 9556f59ad83690640a047192771600600c1550dc. Only changes since last review were suggested address and lock annotation fixes.

I also updated my branch. If any other reviewers struggling with this PR, I'd encourage them to use it to see a breakdown of changes in the main PR:

| [`pr/keyman`](https://github.com/ryanofsky/bitcoin/commits/pr/keyman) | |
| --- | --- |
| 3d38bf3d0c620755ff9ce9a53eb161e2d5ffd055 MOVEONLY: Move key handling code out of wallet to keyman file | +24&nbsp;-6&nbsp;⇔1375 |
| c80a4c7787ec84873e90a101cb4e5a99e4d797f0 Refactor: Split up CWallet and LegacyScriptPubKeyMan and classes | +590&nbsp;-350&nbsp;⇔296 |
| 1529a4bc58035c3f1768830a4a7376086c93b9c6 MOVEONLY: Reorder LegacyScriptPubKeyMan methods | +4&nbsp;-5&nbsp;⇔87 |
| fdc32d358c38e8719f1249d42ab8f2f77d4ce270 Refactor: Declare LegacyScriptPubKeyMan methods as virtual | +21&nbsp;-6&nbsp;⇔1 |
| b4a321dd967124c9c52f68ae4b3e40a1f7861606 Refactor: Add new ScriptPubKeyMan virtual methods | +46&nbsp;-8&nbsp;⇔0 |
| 46fe6e240acfe947ee8100ac9cdfcf9e6234a3ea Refactor: Move SetAddressBook call out of LegacyScriptPubKeyMan::GetNewDestination | +7&nbsp;-4&nbsp;⇔2 |
| 5982116cd8db96e05782ecc5f7825cfd9ec04498 Refactor: Move SetWalletFlag out of LegacyScriptPubKeyMan::UpgradeKeyMetadata | +2&nbsp;-2&nbsp;⇔0 |
| 9b121c57197d96ffdf79740b2b703b2aa89e53f2 Refactor: Remove UnsetWalletFlag call from LegacyScriptPubKeyMan::SetHDSeed | +2&nbsp;-1&nbsp;⇔0 |
| f7d1a7d50b78671e48f269030739d46ed75656d9 Refactor: Move SetAddressBookWithDB call out of LegacyScriptPubKeyMan::ImportScriptPubKeys | +6&nbsp;-4&nbsp;⇔5 |
| 01dd97cb9f15f2ea8d86f86a4bf4be4a39972bbe Refactor: Move LoadKey LegacyScriptPubKeyMan method definition | +6&nbsp;-1&nbsp;⇔0 |
| dda6d03e4969e6b2a71b7f38c602d900c6faca49 Refactor: Move GetMetadata code out of getaddressinfo | +21&nbsp;-8&nbsp;⇔5 |
| 00ee9e9467c66a634700bc134f21d83eb429354b Refactor: Move MarkUnusedAddresses code out of CWallet::AddToWalletIfInvolvingMe | +12&nbsp;-5&nbsp;⇔14 |
| 80b3a0763f643af7d9c9be5ce6b8d6bd3c42938f Refactor: Move Upgrade code out of CWallet::CreateWalletFromFile | +24&nbsp;-12&nbsp;⇔20 |
| f733cbc54693d1a1efffbe0613d53d606c2464ad Refactor: Move HavePrivateKeys code out of CWallet::CreateWalletFromFile | +13&nbsp;-2&nbsp;⇔0 |
| 0d5f789e6fc249dba4125e3d22a7c365b3440789 Refactor: Move SetupGeneration code out of CWallet | +22&nbsp;-11&nbsp;⇔12 |
| 993d1f1a63d0cd837aab06a7dcb03272b5b02f03 Refactor: Move RewriteDB code out of CWallet | +20&nbsp;-4&nbsp;⇔5 |
| 20b68aefd252c4273ea82d672ae1459274c3e48e Refactor: Move GetKeypoolSize code out of CWallet | +12&nbsp;-3&nbsp;⇔8 |
| ee77be668f61a9703e22d15cc14f60e68fb2b4e4 Refactor: Move nTimeFirstKey accesses out of CWallet | +16&nbsp;-3&nbsp;⇔0 |
| 045f7de4f6d7d5e0aa927b726d277dccaee7fb5e Refactor: Move encryption code between KeyMan and Wallet | +67&nbsp;-49&nbsp;⇔24 |
| 6a622f3e94193b5d8da19904a6c2ba7d1fe4b003 Locking: Lock cs_KeyStore instead of cs_main in legacy keyman | +55&nbsp;-73&nbsp;⇔37 |
| f2d0b6bf38f51d7a584db9ba0556981ab5051efc Refactor: Allow LegacyScriptPubKeyMan to be null | +72&nbsp;-28&nbsp;⇔0 |
| 713ff35d756c2c082a8b0bfdd6789f9673bd07ed Refactor: Require scriptPubKey to get wallet SigningProvider | +101&nbsp;-24&nbsp;⇔0 |
| 3e0b0e476e4ebf59d688af009900ed5cb8528221 Key pool: Move CanGetAddresses call | +2&nbsp;-1&nbsp;⇔2 |
| 7126df82214e0d1e5ea47ce43de49b021f359070 Key pool: Move LearnRelated and GetDestination calls | +9&nbsp;-8&nbsp;⇔0 |
| 721b34107984104e1f2b232016e456a7c76a05a6 Key pool: Move TopUp call | +0&nbsp;-1&nbsp;⇔1 |
| 22524983096745e8df808efa57671c86f7d93245 Key pool: Change ReturnDestination interface to take address instead of key | +12&nbsp;-14&nbsp;⇔0 |
| df7f357fb81837edbf0e977ad319af773d10d5af Key pool: Make TopUp fail if unexpected wallet flags are set | +3&nbsp;-0&nbsp;⇔0 |
| 5afee37fdfc10e7cd210e68f3052d1fb1ef658f5 Key pool: Fix omitted pre-split count in GetKeyPoolSize | +1&nbsp;-1&nbsp;⇔0 |
| a71dc57498345e8e506fc52bcb1f6723035f8082 IsMine: Set state to WATCH_ONLY if we can get the pubkey | +6&nbsp;-2&nbsp;⇔0 |
| 6cfb305e01113325ec0811fea8a0104536cbf967 HD Split: Avoid redundant upgrades | +1&nbsp;-1&nbsp;⇔0 |
| 4040cc10d67c87aa942f618213bb1195c7b4db4e Box the wallet: Add multiple keyman maps and loops | +161&nbsp;-46&nbsp;⇔22 |
| 6394d8f92c67f180cb25352a36073cf563315760 Refactor: Copy CWallet signals and print function to LegacyScriptPubKeyMan | +22&nbsp;-16&nbsp;⇔6 |
| 85b944bde87bcc54f0cfc25305dc2d8073a88941 Cleanup: Update message strings and comments | +4&nbsp;-5&nbsp;⇔3 |
| f25201cac1aa2e8d4191786177da9458c79a4e86 Cleanup: Drop unused GUI learnRelatedScripts method | +0&nbsp;-5&nbsp;⇔0 |
| 6da50b7ab6b68c6607e9f9cfc78ea1bf7669d3f0 Refactor: Replace SigningProvider pointers with unique_ptrs | +45&nbsp;-29&nbsp;⇔0 |
| b4be0917a0ccf4fd844a462c4039ce7a760448c4 Refactor: Make LegacyScriptPubKeyMan::HasEncryptedData() call WalletStorage::HasEncryptionKeys() | +7&nbsp;-7&nbsp;⇔0 |
| fa8f61cd3748b3b5fa9b6ffbe33d9262248b127c Refactor: Add storage wrapper functions | +44&nbsp;-33&nbsp;⇔0 |
| d590e6b8d41499928910e13e4d2da9b8fd8a954d Refactor: Make WalletDatabase a shared_ptr instead of a unique_ptr | +28&nbsp;-24&nbsp;⇔0 |

ryanofsky commented:
Code review ACK d938bb4f0208ec629122d0b87f45cad3c8b58089. Only comment and formatting changes since my last review. Thanks for digging up the history!

IIUC, could maybe update the PR description to clarify that the "minor change in p2p behaviour" is a bugfix.

jnewbery commented:
Updated the PR description. Thanks for the review!

ariard commented on doc/release-notes-15437.md:
```diff
@@ -32,3 +32,21 @@ Please use the recommended alternatives if you rely on this deprecated feature:
+  code when a transaction is not accepted to the mempool. They still return the
+  verbal reject reason.
+
+* Log messages that previously reported the REJECT reason when a transaction was
```
Not sure about this statement, at least the reject-reason should be logged for the broadcastTransaction call in SubmitMemoryPoolAndRelay, which ones were you thinking off ?

jnewbery replied:
Sorry, this should say REJECT code, not REJECT reason. I'll update the docs.

ariard reviewed (approved):
ACK d938bb4. Reviewed and tested code, grep'ed to see any reference to reject-code, included in tests, also checked that there weren't any other cases like CACHED_INVALID and BLOCK_MISSING_PREV returning a reject-code of 0.

ryanofsky commented on src/net_processing.cpp:
```diff
@@ -1234,11 +1234,12 @@ void PeerLogicValidation::BlockChecked(const CBlock& block, const CValidationSta
+    // to that peer, maybe punish.
+    if (state.IsInvalid() &&
+        it != mapBlockSource.end() &&
+        State(it->second.first)) {
```
Just a suggestion, but imo it wouldn't be a bad idea to break this commit up to separate the behavior changes from the refactoring changes. It probably says something bad about my own code review practices, but I completely missed this change the first time I first reviewed and ACKed this PR. It seems like this commit could nicely be broken into 3 commits: 1) P2P behavior change 2) RPC and logging changes 3) Refactor to remove reject constants arguments

jnewbery replied:
Good idea. I'll do that. That'll also make it easier to back-port the bugfix if we want to do that.

jnewbery committed:
[logging] Don't log REJECT code when transaction is rejected

Remove the BIP61 REJECT code from error messages and logs when a
transaction is rejected.

BIP61 support was removed from Bitcoin Core in
fa25f43ac5692082dba3f90456c501eb08f1b75c. The REJECT codes will be
removed from the codebase entirely in the following commit.

jnewbery committed:
[validation] Fix REJECT message comments

fjahr reviewed (approved):
ACK d938bb4

Code review and ran and reviewed tests.

Just one comment was probably overlooked and +1 for moving the behavior change into a separate commit.

jnewbery commented on src/validation.h:
```diff
@@ -783,9 +783,6 @@ int32_t ComputeBlockVersion(const CBlockIndex* pindexPrev, const Consensus::Para
  * for transactions, to signal internal conditions. They cannot and should not
```
Thanks. Fixed!

jnewbery commented:
Fixed all @ryanofsky @ariard @fjahr comments. Thanks for the review!

ryanofsky reviewed (approved):
Code review ACK 9075d13153ce06cd59a45644831ecc43126e1e82. Only changes since last review are splitting the main commit and updating comments

ariard commented:
ACK 9075d13, changes since last reviewed are splitting them in separate commits to ease understanding and fix nits

MarcoFalke commented:
It looks like the runtime sanity assert

```cpp
assert (state.GetRejectCode() < REJECT_INTERNAL); // Blocks are never rejected with internal reject codes
```

was transformed into a less fatal if check in 5d08c9c579ba8cc7b684105c6a08263992b08d52 and now removed completely. What is the reason for removing it?

MarcoFalke commented:
Oh, I see we used to punish peers when the reject code was `0`. Post-merge ACK
